### PR TITLE
Refactor DDim's product() and add slice_ddim()

### LINKED ISSUE #2880: Add slice function for dim
能否增加一个对dim的切片函数，类似`SliceDDim({1,2,3}, 1, 3)`：取出[1,3)的dim，即{1,2,3}->{2,3}？
目前RecurrentOp的SegmentInput中，对句子进行拆分时，需要用到这个接口。

### PR DESCRIPTION
**Pull request by Canpio**
1. Refactor DDim's `product()` to make it more efficiently.
2. Add `slice_ddim()`.

fix #2880

### CODE REVIEW

File: paddle/framework/ddim.h [hunk 1]
```diff
@@ -81,6 +81,8 @@ std::vector<int> vectorize(const DDim& ddim);
 
 ssize_t product(const DDim& ddim);
 
+DDim slice_ddim(const DDim& dim, int begin, int end);
```
Comment (luotao1) [inline paddle/framework/ddim.h]: 加下注释吧。

File: paddle/framework/ddim.cc [hunk 2]
```diff
@@ -178,13 +179,56 @@ std::vector<int> vectorize(const DDim& ddim) {
...
+
+  void operator()(const Dim<1>& dim) {
+    PADDLE_ENFORCE(end == 1, "End index in ddim slice is out of bound.");
+    vector.push_back(dim.head);
+  }
+};
+
+DDim slice_ddim(const DDim& dim, int begin, int end) {
+  std::vector<int> vec;
+  vec.reserve(end - begin);
```
Comment (QiJune) [inline paddle/framework/ddim.cc]: Do we have to check the value of begin and end here?
such as 
```
begin <= end;
begin>=1;
```

File: paddle/framework/ddim.cc [hunk 3]
```diff
@@ -178,13 +179,56 @@ std::vector<int> vectorize(const DDim& ddim) {
...
+  }
+};
+
 ssize_t product(const DDim& ddim) {
-  ssize_t result = 1;
-  std::vector<int> v = vectorize(ddim);
-  for (auto i : v) {
-    result *= i;
+  ProductVisitor visitor;
+  return boost::apply_visitor(visitor, ddim);
```
Comment (QiJune) [inline paddle/framework/ddim.cc]: I know that a Dim<D> variable can be passed to a DDim variable.
But, how a DDim varibale  passed to a Dim<D> variable?

Comment (Canpio) [inline paddle/framework/ddim.cc]: See https://github.com/Canpio/Paddle/wiki/Majel:-From-Allocation-to-DArray
In the last section, I show how `DArray` is transformed into `Array` in `boost::apply_visitor`. The principle is also suitable for `DDim` and `Dim`.

File: paddle/framework/ddim.h (hunk 1 above)
Comment (Canpio) [inline paddle/framework/ddim.h]: done

File: paddle/framework/ddim.cc (hunk 2 above)
Comment (Canpio) [inline paddle/framework/ddim.cc]: We do these checks in `SliceVectorizeVisitor`.

```cpp
SliceVectorizeVisitor(std::vector<int>& v, int b, int e)
      : vector(v), begin(b), end(e) {
    PADDLE_ENFORCE(begin < end,
                   "Begin index must be less than end index in ddim slice.");
    PADDLE_ENFORCE(begin >= 0,
                   "Begin index can't be less than zero in ddim slice.");
  }
```

and 

```cpp
void operator()(const Dim<1>& dim) {
    PADDLE_ENFORCE(end == 1, "End index in ddim slice is out of bound.");
    vector.push_back(dim.head);
  }
```

File: paddle/framework/ddim.cc [hunk 4]
```diff
@@ -178,13 +179,56 @@ std::vector<int> vectorize(const DDim& ddim) {
...
+    } else {
+      --begin;
+    }
+    --end;
+    if (end > 0) {
+      this->operator()(dim.tail);
+    }
+  }
+
+  void operator()(const Dim<1>& dim) {
```
Comment (QiJune) [inline paddle/framework/ddim.cc]: This implementation is a little tricky. And all implementation related with dim is tricky.
We may consider remove dim later and just use a small vector.

**Pull request merged**